### ISSUE TITLE
Revisiting Few-sample BERT Fine-tuning

### ISSUE THREAD

**Issue by ymgw55** [ymgw55/MLPapers#171]
# Information
### URL
https://openreview.net/forum?id=cO1IH43yUF
### Authors
[Tianyi Zhang](https://openreview.net/profile?id=~Tianyi_Zhang2), [Felix Wu](https://openreview.net/profile?id=~Felix_Wu1), [Arzoo Katiyar](https://openreview.net/profile?id=~Arzoo_Katiyar1), [Kilian Q Weinberger](https://openreview.net/profile?id=~Kilian_Q_Weinberger1), [Yoav Artzi](https://openreview.net/profile?id=~Yoav_Artzi1)
### Published at
ICLR 2021 Poster

# Reading Motivation
#163 で "re-initialize the top L layers of the pretrained transformer encoder, which has been shown to improve performance and stability during fine-tuning" の根拠として挙げられていた論文.

# Summary

fine-tuning時のBERTの分散表現についての研究. 一般的にデータが少ない時にBERTの学習は不安であることに着目. この不安さを引き起こすいくつかの要因は, 1)biasのかかったgradientの推定をする一般的でない最適化手法が広く使われていること, 2)下流タスクの際, BERTのパラメータの大部分が制限されて適応できないこと, 3)fine-tuningの際に少ないイテレーション回数を事前に決め学習する方法が支流であること, であった. これらの要因について実験を行い, 広く知られている学習過程の不安定さを解消する代替案を確認. これをもとに近年提案された小データで学習を安定化させる手法について見直し, 実験で確認した結果, 著者らの手法の後ではそのような手法の影響は大幅に低下した. 

* BERTAdamはbias補正されていないので普通のAdamを使うべき. huggingface/transformers についてはbertadam から adamw に[修正](https://github.com/huggingface/transformers/blob/b9a768b3ffa80c4c19d024f9f42d5917e7d8109e/docs/source/en/migration.mdx)
* BERTのtopのL(bertlargeでL=1〜6)ブロックをpre-train時と同じ方法で初期化することで下流タスクの精度が上がる.
top のブロックを初期化することで, worst case の seed の性能が改善される. 最適な L はデータ依存.
![Screenshot from 2022-05-02 22-11-58](https://user-images.githubusercontent.com/39965001/166239352-9caac8af-41fa-4699-ab03-5ddd1c9c66a0.png)
各ブロックで初期状態のパラメータとfine-tuning後のパラメータを L2 で比較. 通常のBERTは, top (Block 24) のブロックが大きく変化. 初期化するパラメータを増やすことで変化が小さくなる. ただ L を大きくしすぎると今度は, bottom (Figure 7 では Block 6) のブロックでのパラメータの変化が増す.
![Screenshot from 2022-05-02 22-19-06](https://user-images.githubusercontent.com/39965001/166240271-324a6a6c-869c-4e18-b397-3810d6bcc181.png)

* fine-tuning の step size を増すことで性能を上げることができる. またほとんどのデータセットで top のブロックの初期化をした場合は, そうではない場合と比べ少ないイテレーション回数でbest performance に達した. 
Figure 8 では, 3 epochs = 96 steps

![Screenshot from 2022-05-02 22-25-49](https://user-images.githubusercontent.com/39965001/166241246-a545eea4-067b-410e-bb0f-5a08cf4e9a47.png)

* 近年の小データで学習の不安定さを解消する手法のほとんどは, bertadam 由来の bias を解消する効果があるため,  Adamに修正することでpositive な影響は減少. 効いたのはfine-tuning 前にさらにMNLIでfine-tuningする手法  (Transferring via an Intermediate Task) くらいだった.
